File: apps/worker/src/handlers/topic-handlers.ts
import { modelHandler } from "./model-handler";
import { monitoringHandler } from "./monitoring-handler";

export type KafkaHandler = (params: {
  topic: string;
  partition: number;
  messageId: string | number;
  payload: unknown;
  tx?: unknown;
  db?: unknown;
}) => Promise<void>;

export function getHandlerForTopic(topic: string): KafkaHandler {
  if (topic.startsWith("model.")) return modelHandler;
  if (topic.startsWith("monitoring.")) return monitoringHandler;

  // default fallback: no-op handler
  return async () => undefined;
}

File: apps/backend/src/silo/worker/handlers/topic_handlers.py
from __future__ import annotations

from collections.abc import Awaitable, Callable

from silo.worker.handlers.model import model_handler
from silo.worker.handlers.monitoring import monitoring_handler

KafkaHandler = Callable[..., Awaitable[None]]


def get_handler_for_topic(topic: str) -> KafkaHandler:
    if topic.startswith("model."):
        return model_handler
    if topic.startswith("monitoring."):
        return monitoring_handler

    async def _topic_noop_handler(**kwargs: object) -> None:
        del kwargs
        return None

    _topic_noop_handler.__dict__["__worker_handler_name__"] = topic
    return _topic_noop_handler
